### ISSUE TITLE
Follow useful @testset behavior

### ISSUE THREAD

**Issue by aplavin** [julia-vscode/TestItemRunner.jl#51]
Currently, `@testitem` behaves quite different compared to regular `@testset`. It would be better to follow the latter as possible, at least in some useful ways:
1. Early stopping. One can do `return` in `@testset`, and the testset stops executing. Meanwhile, `return` seems to be a no-op in `@testitem`, quite surprisingly.
2. Local scope. `@testset` run in local scope, while `@testitem` seems to have global scope. This difference manifests itself, for example, with warning for innocuous code like
```julia
o = 1
...
for i in 1:5
    o = i
    ...
end

# warns:
# Warning: Assignment to `o` in soft scope is ambiguous because a global variable by the same name exists: `o` will be treated as a new local.
```

I think both can be solved by wrapping code generated by `@testitem` into `let ... end` blocks.

**Comment by nickrobinson251** [comment]
In [ReTestItems.jl](https://github.com/JuliaTesting/ReTestItems.jl), we think of `@testitem` more like a module than a testset (so 1 applies, but we've found it quite an intuitive change from testset). For 2, we "fix" that by running testitems with REPL-like "[softscope](https://docs.julialang.org/en/v1.10-dev/manual/variables-and-scoping/#on-soft-scope)".

**Comment by NHDaly** [comment]
To confirm: you aren't seeing the warnings you describe inside of a testitem, correct? We fix that via the change that Nick described.

For issue 1, yes, this is a notable departure from testsets, but it's an important one.

A testitem must be _indepent,_ and completely isolated, so that it can be run in isolation from the VSCode interface, or in parallel, as part of runtests. For that reason, you must be able to define things like structs, functions, and modules, inside the testitem, which you cannot do in a testset.

^ This is the fundamental motivation for this difference. We try to highlight this difference in the docstrings. Please let us know if there is more that could be improved there. For example, maybe we can make this motivation more clear. Thanks!

**Comment by davidanthoff** [comment]
At the end of the day `return` is only a valid statement in a function, and `@testitem`s aren't functions, but global code in a module. So one just has to use different control flow constructs if one wants to exit early, e.g. `if` blocks ets.

**Comment by davidanthoff** [comment]
It is valid, but it doesn't do anything outside of a function. The reason it works in testsets is that under the hood that macro puts everything into a function. Test item made a different choice, namely to put everything into a module. That has many benefits: much better isolation, ability to define types etc. But it also means `return` won't do anything in a `@testitem`.

I don't want to modify the semantics of `return` here, right now we have a simple mental model and I want to keep it that way.

**Comment by davidanthoff** [comment]
Ha, I have to admit I'm surprised, didn't expect that :) But the principle for testitems I think is still valid: one should think of them as modules, and I'm fairly certain `return` really doesn't do anything inside a module.

**Comment by stevengj** [comment]
> The reason it works in testsets is that under the hood that macro puts everything into a function

`@testset` just wraps a `let` block around the code, which creates a new scope.  You could still do this inside a module if you want.  (Though then you can't use top-level statements like `import`.)

Alternatively, your macro could just call `REPL.softscope` on the AST to give it function-like "soft" scoping behavior.

**Comment by kenahoo** [comment]
I just wanted to point out that the "warning" above about soft scopes isn't just nuisance chatter, it actually changes the behavior of the code relative to what would happen if the code were executed in other Julia contexts, up to & including producing a compilation error.  I pointed this out in a thread here:  https://discourse.julialang.org/t/soft-scopes-globals-in-testitem-s/119338

The TLDR of it is that the following test will succeed in `pkg> test`, but fail with ```LoadError: UndefVarError: `A` not defined``` when running in VS Code's test pane:

```julia
@testitem "foo" begin
    A = 0
    for i in 1:10
        A += i
    end
    @test A == 55
end
```

I think that's a big enough deal that it should be considered a bug.

**Comment by kenahoo** [comment]
@nickrobinson251 I noticed your comment:

> For 2, we "fix" that by running testitems with REPL-like "[softscope](https://docs.julialang.org/en/v1.10-dev/manual/variables-and-scoping/#on-soft-scope)".

Do you mean it should already be using a softscope that should be fixing this issue?  That's not what I'm seeing, is it only in a pre-release version?

**Comment by davidanthoff** [comment]
`@testitem`s have always used the hard scoping rules. I didn't think much about that at all when I created it, but at the end of the day this is probably a case where one can come up with good arguments for either choice. At this point changing it, though, would break backwards compatibility, so I don't really see that as an option.

@kenahoo the test item framework is distinct from the Julia base testing framework. Test items handle a whole lot of things differently from the base test framework, and I don't see why we should consider those different choices bugs, it's just different design decisions.

**Comment by kenahoo** [comment]
> this is probably a case where one can come up with good arguments for either choice

Are there indeed good arguments why a soft scope should not be used?  I certainly find the above counterintuitive behavior an argument against it, I haven't been able to think of arguments for it.

For backward compatibility considerations - are there any cases where changing to soft scope wouldn't just produce the behavior that people have already inserted workarounds to obtain (e.g. 'let' blocks or avoiding 'for' loops)?  At the very least there would have been a warning a user would have to have seen and understood and accommodated somehow.

**Comment by kenahoo** [comment]
(FYI I hit this independently of [and more recently than] the previous people in this thread, and asked about it at https://discourse.julialang.org/t/soft-scopes-globals-in-testitem-s/119338, which is how I found this thread.  Just putting a backlink here to that conversation.)

**Comment by nickrobinson251** [comment]
i'd love to align ReTestItems.jl with the TestItems provided by/for the VS Code extension (and would happily move ReTestItems.jl to be closer to TestItems.jl in some ways to help achieve that), but fwiw i do think the REPL-like softscope is favourable (and is the behaviour the users (developers) at RAI where I work expected too)

**Comment by davidanthoff** [comment]
> Are there indeed good arguments why a soft scope should not be used?

Good question. The way I've thought about is that my mental model for a test item really is a module, and I really like the simplicity of just being able to say "think of your test item as a module, it behaves exactly in that way" (because it is, of course). And I think whenever one creates a new module, it always has hard scoping rule, right? With soft scoping, we would have to essentially say "think of a test item as a module, but unlike modules it uses soft scoping".

Another reason might be that in general soft scoping seems to be for interactive code in Julia? But hard scoping is the default for code in regular Julia files? And at the end of the day test items are code in Julia files. I think I would almost find it more unintuitive if within one code file different scoping rules apply...

On the backwards compat question, there _is_ code that would behave differently if we were to change it, isn't there? Yes, that might trigger a warning at the moment, but at the end of the day we would be changing the behavior...

**Comment by davidanthoff** [comment]
> I think testitems currently behave differently from regular julia code in functions

YES! Because they are _not_ like code in functions. You can do things like define a type inside a test item. They are modules. What I had in mind is that in an example like this
```julia
module foo 
  # Code A
end

@testitem "Bar" begin
  # Code B
end
```
I would want code A and B to behave the same, because that is the right mental analogy between test items and modules.

**Comment by kenahoo** [comment]
> I would want code A and B to behave the same, because that is the right mental analogy between test items and modules.

That's kind of surprising to me.  I think of modules as things that "define stuff", not actually "do stuff".  I don't generally expect to load a module and have it perform its complete job, I generally load it and *then* I choose what stuff I want it to do, by invoking functions and instantiating types etc.  Whereas, for a test, when I execute its code, it does its full job right away.

Is it the case that if we want test-writers to be able to define types & macros within a `@testitem`, it *must* be a module with no soft scope?  Or would adding `REPL.softscope` give a soft scope that would allow those while also providing the scope experience that most people seem to have expected?

Also, now that we have support for initializers via Test Snippets and Test Modules (thank you thank you for that improvement, by the way!), is it maybe less important that users can define types & macros in their tests?  I'm not sure whether it is or isn't, just thinking that might be germane to the question too.

And it just occurred to me: does using `import` in a `@testitem` require a global/hard scope too, and/or would that be affected by a `REPL.softscope`?  I have been finding it very handy lately to use targeted imports in my tests.

**Comment by nickrobinson251** [comment]
> does using import in a @testitem require a global/hard scope too?

Nope (in ReTestItems.jl we require imports be in the testitem but still use softscope)

**Comment by NHDaly** [comment]
@nickrobinson251 If you're able to easily pull them up, it might be good to have a look at the situations that came up in practice at RAI that convinced us to change this.

**Comment by nickrobinson251** [comment]
These are the test cases we have for softscope in ReTestItems.jl testitems: https://github.com/JuliaTesting/ReTestItems.jl/blob/v1.29.0/test/testfiles/_scope_tests.jl  

(I can try see if i can dig up any "real" examples from when we were converting our test codebase to ReTestItems.jl too).
